proposals/TP3-xds-error-propagation.md

```diff
@@ -0,0 +1,121 @@
+
+Note that while the incremental protocol variants can explicitly indicate to the client that a subscribed resource does not exist, the SotW protocol variants require the client to use a 15-second does-not-exist timeout. This adds an unnecessary delay before the client can react to the error. It has also led to cases where control plane slowness has caused clients to incorrectly react as if a resource does not exist.
+
+### Related Proposals:
+
+* [TP1-xds-transport-next.md](TP1-xds-transport-next.md)
+
+## Proposal
+
+The proposal introduces two new fields to the DiscoveryResponse proto. The first, `resource_errors`, will detail Resource-specific errors and be included in both SotW and Incremental xDS protocols. The second, `removed_resources`, will specifically identify resources that no longer exist, addressing a current gap in the SotW protocol's signaling capabilities. 
+
+```textproto
+// New Proto Message
+// Contains detailed error detail for a single resource
+message ResourceError {
+ ResourceName resource_name = 1;
+
+ google.rpc.Status error_detail = 2;
+}
+
+message DiscoveryResponse {
+
+// The version of the response data.
+string version_info = 1;
+
+….
+…
+
+
+ // The control plane instance that sent the response.
+  config.core.v3.ControlPlane control_plane = 6;
+
+// NEW_FIELD
+// An optional repeated list of error details that the control plane 
+// can provide to notify the client of issues for the resources that 
+// are subscribed.
+// This allows the xDS management server to provide optional 
+// notifications in case of unavailability, permissions errors 
+// without the client having to wait for the `config fetch` timeout.
+repeated ResourceError resource_errors = N;
+...
+}
+
+message DeltaDiscoveryResponse {
+  …
+
+  // NEW FIELD
+  repeated ResourceError resource_errors = N;
+}
+```
+
+### Protocol Behavior
+Clients that support this mechanism will use this additional field to obtain error information for resources that the xDS server couldn’t provide. When this message is received, the xDS client should cancel any resource timers for the indicated resource. The client should then handle the error based on the status code, as follows (using [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) terminology):
+
+  * The following codes will be interpretted by clients as transient failures, meaning that the client MUST continue to use previously cached versions of the resource if it has them: UNAVAILABLE, INTERNAL, UNKNOWN.
+  * The following codes will be interpretted by clients as data errors, which the client MAY handle by dropping any previously cached resource: NOT_FOUND, PERMISSION_DENIED
```

Comment:
This is not a behavior change.

Note that use of "MAY" here, which is intended in the [RFC-2119](https://datatracker.ietf.org/doc/html/rfc2119) sense.  Envoy doesn't have to do this if you don't want to.

gRPC has an option today where it can drop previously cached resources when it sees a resource deletion.  This option already exists, so this isn't introducing anything new.

I don't see any conflict between this and the eventual consistency model.  If the resource comes into existence again later, the client will start using it.  All we're really doing here is not treating "does not exist" any differently than any other value of the resource.

Reply:
As a control plane implementation, I don't think I could responsibly decide to use an error code if it the behavior of it is undefined. I would instead use mechanisms that are clearly defined (which already mostly exists). It doesn't seem like we would want this ambiguity?

Reply:
I think individual data planes are going to handle certain error cases differently than others, as evidenced by the fact that gRPC can be configured to ignore resource deletions and Envoy cannot.  I think all we can do here is document what options data planes are allowed so that control planes can make an informed decision about what codes to send.  If you want to guarantee that the client will always keep using the previously cached version, if any, then there are codes you can use that will guarantee that behavior.

With respect to leaving certain codes undefined, the reasoning was basically that we can't think of any reasons why any of the other codes would actually be used for per-resource errors, and it didn't seem wise to impose semantics for those codes in the absence of concrete use cases.  If a control plane wants to be safe, it can just avoid those codes entirely, and then the issue doesn't arise.

Reply:
My personal take on having these error codes open to interpretation (w.r.t client's behavior) is that it may cause more harm than good. The motivation behind a protocol is to present a well-defined contract, and because the error-code value impact the xDS-related behavior, then they should be properly defined.

Reply:
I understand this concern, but I don't think we can avoid it.  There is already a pre-existing mechanism in gRPC to allow different handling for the does-not-exist case (which today is also used to indicate permission problems), so there are already differences between data planes here.  All we're really doing in this proposal is providing a new way to indicate these cases, and documenting the existing behavior that data planes may be configured to provide.

For some discussion of the reasoning behind gRPC's behavior, see [gRFC A53](https://github.com/grpc/proposal/blob/master/A53-xds-ignore-resource-deletion.md) and my current draft of [gRFC A88](https://github.com/grpc/proposal/pull/466), which describes the changes we're going to make in gRPC to support this xRFC.

I'm open to hearing alternatives here, as long as they don't break existing gRPC functionality.